tensorflow/core/kernels/data/dataset_test_params.h

```diff
@@ -0,0 +1,208 @@
+  test::FillValues<T>(&tensor, input_data);
+  return tensor;
+}
+
+// Creates a vector of tensors with the specified dtype, shape, and values.
+template <typename T>
+std::vector<Tensor> CreateTensors(
+    const TensorShape& shape, const std::vector<gtl::ArraySlice<T>>& values) {
+  std::vector<Tensor> result;
+  result.reserve(values.size());
+  for (auto& value : values) {
+    result.emplace_back(CreateTensor<T>(shape, value));
+  }
+  return result;
+}
+
+enum class DatasetParamsType {
+  Range,
+  Batch,
+  Map,
+};
+
+string ToString(DatasetParamsType type);
+
+class RangeDatasetParams;
+class BatchDatasetParams;
+class MapDatasetParams;
+
+class DatasetParams : public core::RefCounted {
+ public:
+  DatasetParams(DataTypeVector output_dtypes,
+                std::vector<PartialTensorShape> output_shapes, string node_name,
+                DatasetParamsType type);
+
+  ~DatasetParams() {}
+
+  // Returns the dataset inputs as a TensorValue vector.
+  virtual Status MakeInputs(gtl::InlinedVector<TensorValue, 4>* inputs) = 0;
+
+  // Checks if the tensor is a dataset variant tensor.
+  static bool IsDatasetTensor(const Tensor& tensor);
+
+  // Creates a BatchDatasetParams. The ownership is transferred to the caller.
+  BatchDatasetParams* Batch(int64 batch_size, bool drop_remainder,
+                            bool parallel_copy, DataTypeVector output_dtypes,
+                            std::vector<PartialTensorShape> output_shapes,
+                            string node_name);
+  // Creates a MapDatasetParams. The ownership is transferred to the caller.
+  MapDatasetParams* Map(std::vector<Tensor> other_arguments,
+                        FunctionDefHelper::AttrValueWrapper func,
+                        std::vector<FunctionDef> func_lib,
+                        DataTypeVector type_arguments,
+                        DataTypeVector output_dtypes,
+                        std::vector<PartialTensorShape> output_shapes,
+                        bool use_inter_op_parallelism,
+                        bool preserve_cardinality, string node_name);
+
+  DataTypeVector output_dtypes;
+  std::vector<PartialTensorShape> output_shapes;
+  string node_name;
+  string iterator_prefix = kDefaultIteratorPrefix;
+  DatasetParamsType type;
+};
+
+class SourceDatasetParams : public DatasetParams {
+ public:
+  SourceDatasetParams(DataTypeVector output_dtypes,
+                      std::vector<PartialTensorShape> output_shapes,
+                      string node_name, DatasetParamsType type);
+
+  ~SourceDatasetParams() {}
+};
+
+class UnaryDatasetParams : public DatasetParams {
+ public:
+  UnaryDatasetParams(DatasetParams* input_dataset_params,
+                     DataTypeVector output_dtypes,
+                     std::vector<PartialTensorShape> output_shapes,
+                     string node_name, DatasetParamsType type);
+
+  ~UnaryDatasetParams() {
+    if (input_dataset_params) {
+      input_dataset_params->Unref();
```

Comment:
Handles the release of input dataset params in the data pipeline.

Reply:
seeing as the ownership of this params object is never shared, I don't think you need refcounting or smart pointer and simply passing / moving the object values around will suffice